File: 0347-39-3160-find-the-number-of-distinct-colors-among-the-balls/0347-39-3160-find-the-number-of-distinct-colors-among-the-balls.py
class Solution:
    def queryResults(self, limit, queries):
        n = len(queries)
        result = [0] * n
        ball_color_map = {}
        color_count = {}

        for i, (ball, color) in enumerate(queries):
            if ball in ball_color_map:
                old_color = ball_color_map[ball]
                color_count[old_color] -= 1
                if color_count[old_color] == 0:
                    del color_count[old_color]

            color_count[color] = color_count.get(color, 0) + 1
            ball_color_map[ball] = color
            result[i] = len(color_count)

        return result

File: 0347-39-3160-find-the-number-of-distinct-colors-among-the-balls/0347-39-3160-find-the-number-of-distinct-colors-among-the-balls.java
import java.util.*;

class Solution {
    public int[] queryResults(int limit, int[][] queries) {
        int n = queries.length;
        int[] result = new int[n];
        Map<Integer, Integer> ballColorMap = new HashMap<>();
        Map<Integer, Integer> colorCount = new HashMap<>();

        for (int i = 0; i < n; i++) {
            int ball = queries[i][0], color = queries[i][1];

            if (ballColorMap.containsKey(ball)) {
                int oldColor = ballColorMap.get(ball);
                colorCount.put(oldColor, colorCount.get(oldColor) - 1);
                if (colorCount.get(oldColor) == 0) {
                    colorCount.remove(oldColor);
                }
            }

            colorCount.put(color, colorCount.getOrDefault(color, 0) + 1);
            ballColorMap.put(ball, color);
            result[i] = colorCount.size();
        }

        return result;
    }
}
